test/utils/mpi/executor.hpp

```diff
@@ -46,59 +46,105 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <ginkgo/core/base/mpi.hpp>
 
 
-void init_executor(std::shared_ptr<gko::ReferenceExecutor> ref,
-                   std::shared_ptr<gko::ReferenceExecutor>& exec)
+template <typename ExecType>
+std::shared_ptr<ExecType> init_executor(
+    std::shared_ptr<gko::ReferenceExecutor>);
+
+
+template <>
+inline std::shared_ptr<gko::ReferenceExecutor>
+    init_executor<gko::ReferenceExecutor>(
+        std::shared_ptr<gko::ReferenceExecutor>)
 {
-    exec = gko::ReferenceExecutor::create();
+    return gko::ReferenceExecutor::create();
 }
 
 
-void init_executor(std::shared_ptr<gko::ReferenceExecutor> ref,
-                   std::shared_ptr<gko::OmpExecutor>& exec)
+template <>
+inline std::shared_ptr<gko::OmpExecutor> init_executor<gko::OmpExecutor>(
+    std::shared_ptr<gko::ReferenceExecutor>)
 {
-    exec = gko::OmpExecutor::create();
+    return gko::OmpExecutor::create();
 }
 
 
-void init_executor(std::shared_ptr<gko::ReferenceExecutor> ref,
-                   std::shared_ptr<gko::CudaExecutor>& exec)
+template <>
+inline std::shared_ptr<gko::CudaExecutor> init_executor<gko::CudaExecutor>(
+    std::shared_ptr<gko::ReferenceExecutor> ref)
 {
-    ASSERT_GT(gko::CudaExecutor::get_num_devices(), 0);
-    exec = gko::CudaExecutor::create(
-        gko::mpi::map_rank_to_device_id(MPI_COMM_WORLD,
-                                        gko::CudaExecutor::get_num_devices()),
-        ref);
+    {
+        if (gko::CudaExecutor::get_num_devices() == 0) {
+            throw std::runtime_error{"No suitable CUDA devices"};
+        }
+        return gko::CudaExecutor::create(
+            gko::mpi::map_rank_to_device_id(
+                MPI_COMM_WORLD, gko::CudaExecutor::get_num_devices()),
+            ref);
+    }
 }
 
 
-void init_executor(std::shared_ptr<gko::ReferenceExecutor> ref,
-                   std::shared_ptr<gko::HipExecutor>& exec)
+template <>
+inline std::shared_ptr<gko::HipExecutor> init_executor<gko::HipExecutor>(
+    std::shared_ptr<gko::ReferenceExecutor> ref)
 {
-    ASSERT_GT(gko::HipExecutor::get_num_devices(), 0);
-    exec = gko::HipExecutor::create(
+    if (gko::HipExecutor::get_num_devices() == 0) {
+        throw std::runtime_error{"No suitable HIP devices"};
+    }
```

Comment:
Is not it handled by `ASSERT_GT`?

Reply:
We call the init functions in the constructor, and I don't think we should use these assertions in the constructor https://google.github.io/googletest/faq.html#CtorVsSetUp. Also the error message is a bit nicer